### DISCUSSION TITLE
.vscode directory

### DISCUSSION THREAD

**Discussion by trampas** [eclipse-theia/theia#14666] [General]
I noticed that theia was using the .vscode directory for configurations.  I found this a bit annoying as that I wanted to test same project with theia and vscode.  However theia did not like the launch.json used for cortex debug that vscode used.

**Comment by msujew** [comment]
Hey @trampas,

> I noticed that theia was using the .vscode directory for configurations. I found this a bit annoying as that I wanted to test same project with theia and vscode.

Theia own its own generally uses the `.theia` directory, but will also use the `.vscode` directory if available. That's usually a convenience feature for most people, but I get why it's annoying in your specific case.

>  However theia did not like the launch.json used for cortex debug that vscode used.

What exactly is Theia complaining about?

**Reply by trampas** [reply]
I was trying to use the cdt_gdb extension which used different format for launch.json.

**Reply by msujew** [reply]
Alright, so did you try using a cdtgdb debug configuration for that?

**Reply by trampas** [reply]
When I found that cdt-gdb was trying to modify the lauch.json that I had setup in vscode for cortex-debug,  I closed and exited theia! 
That is I did not want to risk messing up my vscode configuration to test gdb-cdt. 

Why would theia use the same .vscode directory instead of a  .theia directory? 
Basically felt like theia is nothing more than an extension for vscode and at worse a 'me too' clone of vscode.  At this point my experimentation with theia stopped. 

I have a rule where if I am buying a tool,  if the tool compares it self to another tool, then you buy the other one.  For example if you are buying a Ryobi drill and it compares its self to Dewalt, buy the Dewalt.  As such theia was seemed to be comparing its self to vscode, so I should just use vscode.   Even worse theia was basically saying "I am just a clone of vscode in a new skin". 

Note I was looking at theia because I disliked how vscode for embedded development felt like gluing together countless extension hacks.  That is you have to click on the "cortex-debug" extension to load and debug target.  This also means the "debug" button does not save an build project.  The same with the vscode-makefile extension.  Here you have to go to the extension and click on the build button, does not even allow keyboard short cuts.   As such I hoped that theia might be better integrated, like eclipse.  However it suffers the same as vscode and leaves you feeling dirty like, why use a knock off if it sucks as bad as the original?   

My conclusion was that both vscode and theia suck for embedded development at this point in time.  Theia sucked worse as it was just a vscode me too skin of vscode.   I also realized that the only way to use vscode/theia for embedded was for me to write an extension.  
Then I realized vscode extensions is the javascript framework hell being transposed on embedded development.  That is there is N javascript frameworks and they all suck, so lets write a new framework that takes the best of the N frameworks.  Now we have N+1 javascript frameworks with issues. 

Now my first impressions could be totally wrong, if so then consider why and what could theia do to improve this?

**Reply by msujew** [reply]
Thanks for the feedback @trampas. The reason why Theia might seem like a worse VS Code is twofold:

1. Historically, Theia was - and still very much is - a platform for building IDE tools, similar to the original Eclipse RCP system. That's where the embedded part of your post comes in. In the past years, a lot of microchip/hardware producers have opted to build IDEs based on Theia for their specific hardware. See the [Arduino IDE](https://github.com/arduino/arduino-ide), [ARM's Mbed studio](https://os.mbed.com/studio/) or [CCStudio by Texas Instruments](https://www.ti.com/tool/CCSTUDIO). We just used to build the base for these companies to built their own tools. Theia as a platform was therefore never designed to cater to in-depth requirements for specific use cases.
2. We recently (only a few months ago) came out with the Theia IDE, which is essentially just the Theia platform with a few goodies on top. So there's definitely still a lot of room for improvement. However, given our lack of resources (compared to a company like Microsoft), our "product" cannot and will likely never be as good as the original. The reason for offering the Theia IDE anyway is for people who don't want to use Microsoft's VS Code and their closed source ecosystem, see [this fantastic blog post on the topic](https://ghuntley.com/fracture/). It also acts as a funnel for people to get to know the platform (see point 1) so maybe they or the company they work for start building their own IDE based on our framework.

Embedded development often times requires very specific tooling for very specific purposes (especially when it comes to the area of compiling & debugging). Both VS Code and stock Theia are very much general purpose tools (which work great for languages like JavaScript or Python), but distinctly lack in other areas. Our attempt at "fixing" this is to provide the framework for others to built upon, so they can built a specifically crafted IDE for embedded development.

**Comment by trampas** [comment]
Mark, 

Thanks for the information! 

For myself I do not like the vendor tools for embedded as they often have too many bugs and issues.  For example I will 'no bid' a contract if they used MPLABX, as there is no way to estimate the number of times during the day the IDE will crash other than to know it is greater than 10. 
I have typically used eclipse with just the cdt to do embedded development.   This works great as at the end of the day the only things needed for embedded is the configuring the cross compiler (which is just path to compiler) and then how to load code on target and start the gdb server. Here eclipse handles both very well out of the box using cdt. 

Eclipse is a little messed up with setting the properties(compile options), paths, etc.  Here I would be happier to have cmake file or makefile, but once you have mastered the properties in eclipse it is relatively easy use and is consistent.  Generally I find I do this once and forget, so it is easy for other developers on team as hey never need to look at the properties to add code or fix bugs. 

The bad part of eclipse is that it is bloat ware, it does not have the gitlens feature of vscode, and hard for newbie to install and configure everything such that they can clone code from git and build the project.  That is it takes too many independent programs to install, configure paths, etc.   I am not sure how vendors are doing this with theia, like how they install segger jlink tools, compiler, etc? 

My point is that many of the features needed for vendor specific IDEs using theia is the base foundation of how to build project and load on hardware and debug with jlink.   If theia had this native or with one plug in I would be excited.    I know every vendor will want to add their own GUI code generator, pin assignment GUI and junk that I personally never use.  However the base build and debug layer would be good to have in theia independently.  

For example, with cortex-debug, if I make a change to source code cortex-debug does not know the project needs rebuilt before debugging.  This is hard for cortex-debug to handle because it does not know which plugin is doing the builds.  Here again a good plug-in that combines all these items is needed. 

Note I would love it if the desktop version also installed xpack.  I love xpack to install development tools!  I would love it if the build tool could do the 'xpm install' if the tool in xpacks directory is missing.   

However it seems most people are going to a container development environment setup, where they use docker setup and use same docker setup for the CI/CD using github or other git server setup.
